src/CodeAnalysis/CSharp/MetaAnalyzers/CSharpImmutableObjectMethodAnalyzer.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Analyzers;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Microsoft.CodeAnalysis.CSharp.Analyzers.MetaAnalyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class CSharpImmutableObjectMethodAnalyzer : DiagnosticAnalyzer
+    {
+        // Each analyzer needs a public id to identify each DiagnosticDescriptor and subsequently fix diagnostics in CodeFixProvider.cs
+        private static readonly LocalizableString s_localizableTitle = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocationTitle), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocationMessage), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocationDescription), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+
+        public static DiagnosticDescriptor DoNotIgnoreReturnValueDiagnosticRule = new DiagnosticDescriptor(
+            DiagnosticIds.DoNotIgnoreReturnValueOnImmutableObjectMethodInvocation,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.AnalyzerCorrectness,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            description: s_localizableDescription,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+        {
+            get
+            {
+                return ImmutableArray.Create(DoNotIgnoreReturnValueDiagnosticRule);
+            }
+        }
+
+        private static readonly string s_solutionFullName = @"Microsoft.CodeAnalysis.Solution";
+        private static readonly string s_projectFullName = @"Microsoft.CodeAnalysis.Project";
+        private static readonly string s_documentFullName = @"Microsoft.CodeAnalysis.Document";
+        private static readonly string s_syntaxNodeFullName = @"Microsoft.CodeAnalysis.SyntaxNode";
+
+        // A list of known immutable object names
+        private static ImmutableArray<string> s_immutableObjectNames
```

Comment:
@tmeschter Thoughts on ImmutableArray here? I originally tried 

```CSharp
private static readonly ImmutableArray<string> s_immutableObjectNames = new ImmutableArray<string>()
{
	s_solutionFullName,
	s_projectFullName,
	s_documentFullName,
	s_syntaxNodeFullName,
};
```

However while this compiled, I think it ends up calling `.Add()` on a null object and I was getting `TypeLoadExceptions`.

Creating the array every time it's requested seems strange too, and I didn't find anything in the coding guidelines about private properties. Let me know if this looks OK.